### PR TITLE
[Fix #556] Show deleted stories and comments to moderators

### LINKED ISSUE #556: show deleted stories/comments to mods
Few stories get deleted, but it's helpful for mods to see them. For example, [this outright spammer](
https://lobste.rs/u/pradm)'s profile shows 1 story submitted, but it was deleted. It should show to mods on [his /newest page](https://lobste.rs/newest/pradm). Similarly, via a [sockpuppet](https://lobste.rs/u/ceopsm) he posted a comment and it'd be helpful if it appeared on [that account's /comments page](https://lobste.rs/threads/ceopsm).

Also it'd be nice if the profile showed the correct stories count based on `@user.is_moderator?` - right now story count includes deleted for both and comment count does not include deleted for either. It'd be twice as nice if it showed both counts to mods, like "0 (+1 deleted)".

### PR DESCRIPTION
**Pull request by jmks**
A fix for #556: Moderators can now see deleted stories (at `/newest/:username`) and comments (at `/threads/:username`). Additionally, on the profile page it now shows as: 
screen shot 2018-10-08 at 3 06 33 pm

Regular users now only see the count as the number of stories or comments they would see if they followed the link.

I tried to follow existing patterns. Notably, this included adding new keys to the `Keystore`, specifically `user:user_id:stories_deleted` and `user:user_id:comments_deleted`, to display the counts on the user's profile page.

### CODE REVIEW

**Comment by jmks** [comment]
Re: CI, `master` is currently failing. I'll wait till #564 or similar is merged to rebase.

**Comment by pushcx** [comment]
#564 is merged. Thanks for taking a crack at this; I plan to review on Wednesday morning.

File: db/migrate/20181003234523_initialize_stories_deleted_counts.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class InitializeStoriesDeletedCounts < ActiveRecord::Migration[5.2]
+  def up
+    Keystore.transaction do
```
Comment (pushcx) [inline db/migrate/20181003234523_initialize_stories_deleted_counts.rb]:
```suggestion
    ActiveRecord::Base.connection.execute <<~SQL
      replace low_priority into keystores
      select concat("user:", users.id, ":stories_deleted"), 0
      from users left join stories on users.id = stories.user_id where stories.id is null;
      
      replace low_priority into keystores
      select concat("user:", users.id, ":stories_deleted"),  count(*)
      from users left join stories on users.id = stories.user_id
      where stories.is_expired = true group by users.id;
    SQL
```

Naming an ActiveRecord model in a migration is a setup for future frustration - if the model is removed or renamed, the migration can't run. I'll add the `good_migrations` gem to catch this, but for now here's a way to both avoid that and not only avoid (10k users * 3) round trips but make the database do all the work.

Please merge the other migration into this one to reuse this strategy.

File: app/helpers/users_helper.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
+module UsersHelper
```
Comment (pushcx) [inline app/helpers/users_helper.rb]: Good call to break this complexity out into a helper.

**Comment by jmks** [comment]
Thanks! Sorry for the delay, I'll finish this up tomorrow.

**Comment by jmks** [comment]
@pushcx I've merged the migrations to use only SQL and merged them.
However, I'm having trouble on the `down` migration. The `key` column collides with a reserved word, so it needs to be escaped.

The original migration used `arel` and worked fine:

```
2.3.3 :001 > condition = Keystore.arel_table[:key].matches("user:%:stories_deleted")
2.3.3 :002 > Keystore.where(condition).delete_all

# (12.2ms)  SET NAMES utf8,  @@SESSION.sql_mode = CONCAT(CONCAT(@@sql_mode, ',STRICT_ALL_TABLES'), ',NO_AUTO_VALUE_ON_ZERO'),  @@SESSION.sql_auto_is_null = 0, @@SESSION.wait_timeout = 2147483
# Keystore Destroy (0.3ms)  DELETE FROM `keystores` WHERE `keystores`.`key` LIKE 'user:%:stories_deleted'
#=> 0
```

When I run that same query in the DB or Rails console it seems to work fine, but when I attempt to run the down migration, it always fails for me:

```
$ rake db:migrate:down VERSION=20181003234523
== 20181003234523 InitializeStoriesDeletedAndCommentsDeletedCounts: reverting =
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

Mysql2::Error: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'DELETE FROM `keystores` WHERE `keystores`.`key` LIKE 'user:%:comments_deleted'' at line 2: DELETE FROM `keystores` WHERE `keystores`.`key` LIKE 'user:%:stories_deleted'
DELETE FROM `keystores` WHERE `keystores`.`key` LIKE 'user:%:comments_deleted'
HOME/.rvm/gems/ruby-2.3.3/gems/mysql2-0.5.2/lib/mysql2/client.rb:131:in `_query'
HOME/.rvm/gems/ruby-2.3.3/gems/mysql2-0.5.2/lib/mysql2/client.rb:131:in `block in query'
HOME/.rvm/gems/ruby-2.3.3/gems/mysql2-0.5.2/lib/mysql2/client.rb:130:in `handle_interrupt'
HOME/.rvm/gems/ruby-2.3.3/gems/mysql2-0.5.2/lib/mysql2/client.rb:130:in `query'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:187:in `block (2 levels) in execute'
HOME/.rvm/gems/ruby-2.3.3/gems/activesupport-5.2.1/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
HOME/.rvm/gems/ruby-2.3.3/gems/activesupport-5.2.1/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
HOME/.rvm/gems/ruby-2.3.3/gems/activesupport-5.2.1/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:186:in `block in execute'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract_adapter.rb:579:in `block (2 levels) in log'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract_adapter.rb:578:in `block in log'
HOME/.rvm/gems/ruby-2.3.3/gems/activesupport-5.2.1/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract_adapter.rb:569:in `log'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:185:in `execute'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/connection_adapters/mysql/database_statements.rb:28:in `execute'
HOME/code/lobsters/db/migrate/20181003234523_initialize_stories_deleted_and_comments_deleted_counts.rb:27:in `down'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:817:in `exec_migration'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:798:in `block (2 levels) in migrate'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:797:in `block in migrate'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:414:in `with_connection'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:796:in `migrate'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:977:in `migrate'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:1292:in `block in execute_migration_in_transaction'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:1345:in `ddl_transaction'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:1291:in `execute_migration_in_transaction'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:1250:in `run_without_lock'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:1202:in `block in run'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:1363:in `with_advisory_lock'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:1202:in `run'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/migration.rb:1050:in `run'
HOME/.rvm/gems/ruby-2.3.3/gems/activerecord-5.2.1/lib/active_record/railties/databases.rake:115:in `block (3 levels) in <top (required)>'
HOME/.rvm/gems/ruby-2.3.3/gems/rake-12.3.1/exe/rake:27:in `<top (required)>'
HOME/.rvm/gems/ruby-2.3.3/bin/ruby_executable_hooks:15:in `eval'
HOME/.rvm/gems/ruby-2.3.3/bin/ruby_executable_hooks:15:in `<main>'
 ...
```

I rarely use mysql but Rails seems to be escaping the columns correctly:

```
2.3.3 :004 > ActiveRecord::Base.connection.quote_column_name "key"
# => "`key`"
```

The only other migrations for the `keystores` table use the `arel`-solution (which I copied from). Any ideas?

**Comment by pushcx** [comment]
I pulled your branch and and sorted it by making it ActiveRecord's problem. I'm going to locally squash and merge this work, so on GitHub this PR may not show as merged. Let's see...

**Comment by pushcx** [comment]
Merged in 18167b36a5b01644932c826d5833ea0b7cd1daeb, deploying in a bit with some other changes.

Thanks for this, you've nailed a confusing bug while improving usability.

**Comment by jmks** [comment]
Thanks for the help finishing this up!

**Pull request closed without merging**